### PR TITLE
Replace BTreeSet of incompleted block with first block logic

### LINKED ISSUE #1333: Optimise `Scheduler`'s `MissedBlocks` by replacing with minimal value
### File Location(s)

`common/scheduler`
`pallets/gear-scheduler`

### Proposal

For now, I think it's better to keep just the smallest block number, which still has some tasks to do, going from that one till the current number. Atm we store set of this numbers.

### PR DESCRIPTION
**Pull request by MikitaS**
Signed-off-by: MikitaS <mikita.stryzhak@gmail.com>

Resolves #1333

### CODE REVIEW

**Comment by MikitaS** [comment]
@breathx PTAL

File: pallets/gear/src/lib.rs [hunk 1]
```diff
@@ -935,40 +935,41 @@ pub mod pallet {
...
-            for bn in &missed_blocks {
-                // Tasks drain iterator.
-                let tasks = TaskPoolOf::<T>::drain_prefix_keys(*bn);
+            // Avoiding iterators for `T::BlockNumber`.
+            let first_block_int: u32 = first_missed_block.unique_saturated_into();
+            let last_block_int: u32 = bn + 1;
 
+            // Iterating over all blocks.
+            for process_bn in first_block_int..last_block_int {
+                let process_bn_v : T::BlockNumber = process_bn.into();
```
Comment (shamilsan) [inline pallets/gear/src/lib.rs:963]:
```suggestion
                let process_bn_v: T::BlockNumber = process_bn.into();
```

File: pallets/gear/src/lib.rs [hunk 2]
```diff
@@ -935,40 +935,41 @@ pub mod pallet {
...
-            // Taking block numbers, where some incomplete tasks held.
-            // If there are no such values, we charge for single read, because
-            // nothing changing in database, otherwise we delete previous
-            // value and charge for single write.
-            //
-            // We also append current bn to process it together, by iterating
-            // over sorted bns set (that's the reason why `BTreeSet` used).
-            let (missed_blocks, were_empty) = MissedBlocksOf::<T>::take()
-                .map(|mut set| {
+            let bn : u32 = <frame_system::Pallet<T>>::block_number().unique_saturated_into();
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply pallets/gear/src/lib.rs]: done

File: pallets/gear/src/lib.rs (hunk 1 above)
Comment (MikitaS) [reply pallets/gear/src/lib.rs]: done

File: pallets/gear/src/lib.rs [hunk 3]
```diff
@@ -935,40 +935,41 @@ pub mod pallet {
...
+
+            // Taking first block number, those has some incomplete tasks held.
+            // If there are no such blocks, we charge for single read, because
+            // nothing changing in database
+            let (first_missed_block, were_empty) = MissedBlocksOf::<T>::take()
+                .map(|first_bn| {
+                    // TODO: replase write with read
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply pallets/gear/src/lib.rs]: this TODO delereted 'cause of not actual

**Comment by breathx** [comment]
also simplify algo of processing tasks, please, if it possible

File: pallets/gear/src/lib.rs [hunk 4]
```diff
@@ -935,40 +935,41 @@ pub mod pallet {
...
-            let (missed_blocks, were_empty) = MissedBlocksOf::<T>::take()
-                .map(|mut set| {
+            let bn: u32 = <frame_system::Pallet<T>>::block_number().unique_saturated_into();
+
+            // Taking first block number, those has some incomplete tasks held.
+            // If there are no such blocks, we charge for single read, because
+            // nothing changing in database
+            let (first_missed_block, were_empty) = MissedBlocksOf::<T>::take()
+                .map(|first_bn| {
+                    // TODO: replase write with read
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply pallets/gear/src/lib.rs]: it's deleted

File: pallets/gear/src/lib.rs [hunk 5]
```diff
@@ -935,40 +935,40 @@ pub mod pallet {
...
             // When we had to stop processing due to insufficient gas allowance.
             let mut stopped_at = None;
 
-            // Iterating over blocks.
-            for bn in &missed_blocks {
-                // Tasks drain iterator.
-                let tasks = TaskPoolOf::<T>::drain_prefix_keys(*bn);
+            // Avoiding iterators for `T::BlockNumber`.
+            let first_block_int: u32 = first_missed_block.unique_saturated_into();
+            let last_block_int: u32 = bn + 1;
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply pallets/gear/src/lib.rs:958]: done

File: pallets/gear/src/lib.rs [hunk 6]
```diff
@@ -935,40 +935,41 @@ pub mod pallet {
...
-            // We also append current bn to process it together, by iterating
-            // over sorted bns set (that's the reason why `BTreeSet` used).
-            let (missed_blocks, were_empty) = MissedBlocksOf::<T>::take()
-                .map(|mut set| {
+            let bn: u32 = <frame_system::Pallet<T>>::block_number().unique_saturated_into();
+
+            // Taking first block number, those has some incomplete tasks held.
+            // If there are no such blocks, we charge for single read, because
+            // nothing changing in database
+            let (first_missed_block, were_empty) = MissedBlocksOf::<T>::take()
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply pallets/gear/src/lib.rs:943]: done

File: pallets/gear/src/lib.rs [hunk 7]
```diff
@@ -935,40 +935,41 @@ pub mod pallet {
...
-            // Taking block numbers, where some incomplete tasks held.
-            // If there are no such values, we charge for single read, because
-            // nothing changing in database, otherwise we delete previous
-            // value and charge for single write.
-            //
-            // We also append current bn to process it together, by iterating
-            // over sorted bns set (that's the reason why `BTreeSet` used).
-            let (missed_blocks, were_empty) = MissedBlocksOf::<T>::take()
-                .map(|mut set| {
+            let bn: u32 = <frame_system::Pallet<T>>::block_number().unique_saturated_into();
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply pallets/gear/src/lib.rs:938]: done

File: pallets/gear-scheduler/src/lib.rs [hunk 8]
```diff
@@ -129,7 +129,7 @@ pub mod pallet {
     /// Defines block number, which should already contain no tasks,
     /// because they were processed before.
     /// Missed blocks processing prioritized.
-    type MissedBlocksCollection<T> = BTreeSet<BlockNumberFor<T>>;
+    type MissedBlocksCollection<T> = BlockNumberFor<T>;
```
[the remark this answers is not in the document]
Comment (MikitaS) [reply pallets/gear-scheduler/src/lib.rs:132]: done

**Comment by MikitaS** [comment]
> also simplify algo of processing tasks, please, if it possible

Seems that it's already simplified. No any functional simplification for performance or readability

**Comment by breathx** [comment]
CI fixes ?

**Comment by ark0f** [comment]
Closing in favor of #2404

**Pull request closed without merging**